qcfractal/interface/collections/collection.py

```diff
@@ -0,0 +1,164 @@
+        created.
+
+        Parameters
+        ----------
+        name : str
+            The name of the Collection object as ID'ed on the storage backend@
+        client : client.FractalClient, optional
+            A Portal client to connect to a server
+        **kwargs
+            Additional keywords which are passed to the Collection and the initial data constructor
+            It is up to the individual implementations of the Collection to do things with that data
+        """
+
+        # Client and mongod objects
+        self.client = client
+
+        # Blank data object
+        self.data = {
+            "name": name,
+            "collection": self.__class__.__name__,
+            "collection_index": (self.__class__.__name__, name),
+            "provenance": {},
+            **self._init_collection_data(kwargs)
+        }
+
+        # Index and internal data
+        self.df = pd.DataFrame()
+
+        if self.client is not None:
+
+            if not isinstance(client, FractalClient):
+                raise TypeError("Storage: client argument of unrecognized type '{}'".format(type(client)))
+
+            tmp_data = self.client.get_collections([self.data["collection_index"]])
+            if len(tmp_data) == 0:
+                print("Warning! {} `{}: {}` not found, creating blank database.".format(
+                    self.__class__.__name__, *self.data["collection_index"]))
+            else:
+                # Augment data with extra fields which may have come from the Collection itself
+                self.data = {**self.data, **tmp_data[0]}
+
+    @abc.abstractmethod
+    def _init_collection_data(self, additional_data_dict):
+        """
+        Additional data defined by the Collection
+
+        This is in addition to the default data. If there is no additional data, simply return and empty dict
+
+        Parameters
+        ----------
+        additional_data_dict : dict
+            Additional data which the individual implementation can work with
+
+        Returns
+        -------
+        collection_data : dict
+            Additional data to be added as part of the collection.
+            If the collection does not have additional data, return this as an empty dict
+        """
+        raise NotImplementedError()
+
+    def to_json(self, filename=None):
+        """
+        If a filename is provided, dumps the file to disk. Otherwise returns a copy of the current data.
+
+        Parameters
+        ----------
+        filename : str, Optional, Default: None
+            The filename to drop the data to.
+
+        Returns
+        -------
+        ret : dict
+            A JSON representation of the Database
+        """
+        if filename is not None:
+            with open(filename, 'w') as open_file:
+                json.dump(self.data, open_file)
+
+        else:
+            return copy.deepcopy(self.data)
+
+    def _pre_save_prep(self, client):
```

Comment:
I did not make this an abstract class because it is not required to be implemented in the subclasses. The basic class just does a `pass` and I don't know if I want to make people implement a method which may simply just `pass`